### PR TITLE
Added sign-in component

### PR DESCRIPTION
**Pull request by IvanShterev**

### CODE REVIEW

File: express-server/src/app.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ const express = require('express');
 const cors = require('cors');
 const app = express();
 const bodyParser = require('body-parser');
-const originsWhitelist = 'http://localhost:4200';
+const originsWhitelist = 'http://localhost:8100';
```
Comment (ochikov) [inline express-server/src/app.js]: Consider moving it in env file

File: ionic-angular-template/src/auth/sign-in/sign-in.component.html [hunk 2]
```diff
@@ -0,0 +1,68 @@
+<ion-header>
+  <ion-toolbar>
+    <ion-title>Login</ion-title>
+  </ion-toolbar>
+</ion-header>
+
+<ion-content class="ion-padding">
+  <form #f="ngForm" (ngSubmit)="onSubmit(f)">
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.html]: Consider using reactive forms in Angular, not template driven forms :)

File: ionic-angular-template/src/auth/sign-in/sign-in.component.scss [hunk 3]
```diff
@@ -0,0 +1,7 @@
+form {
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.scss]: you can use ionic classes: "ion-padding-start ion-padding-end ion-padding-top ion-padding-bottom" as well as margins and etc. 
Please review: https://ionicframework.com/docs/layout/css-utilities

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  constructor(
+    private authService: AuthService,
+    private router: Router,
+    private alertCtrl: AlertController
+  ) { }
+
+  ngOnInit() {}
+
+  onSubmit(form: NgForm) {
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: Consider using reactive forms, not template driven forms

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  public password: FormControl;
+  constructor(
+    private authService: AuthService,
+    private router: Router,
+    private alertCtrl: AlertController,
+    private fb: FormBuilder
+  ) { }
+
+  ngOnInit() {
+    this.email = new FormControl('', Validators.compose([Validators.required, Validators.email]));
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: These are already part of the form group. There is no need to add them separately. You can move the initial value, as well as the Validators in the creation of the form group.

File: ionic-angular-template/src/auth/sign-in/sign-in.component.html [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            <ion-item>
+              <ion-label position="floating">E-Mail</ion-label>
+              <ion-input
+                type="email"
+                formControlName='email'
+                name="email"
+              ></ion-input>
+            </ion-item>
+            <ion-item
+              *ngIf="!email.valid && email.touched"
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.html]: !email.valid = email.invalid

File: ionic-angular-template/src/auth/sign-in/sign-in.component.html [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            <ion-item class="item">
+              <ion-label position="floating">Password</ion-label>
+              <ion-input
+                type="password"
+                formControlName='password'
+                name="password"
+              ></ion-input>
+            </ion-item>
+            <ion-item
+              *ngIf="!password.valid && password.touched"
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.html]: !password.valid = password.invalid

File: ionic-angular-template/src/auth/sign-in/sign-in.component.html (hunk 6 above)
Comment (ochikov) [reply ionic-angular-template/src/auth/sign-in/sign-in.component.html]: Also let's check first if it is touched and then valid or invalid ?

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    this.password = new FormControl('', Validators.compose([Validators.required, Validators.minLength(6)]));
+    this.signInForm = this.fb.group({
+      email: this.email,
+      password: this.password
+    });
+  }
+
+  async signIn(email: string, password: string) {
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: You can get the values of the form from the getters. You can implement getter for every formControlName.

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+})
+export class SignInComponent implements OnInit {
+  public signInForm: FormGroup;
+  public email: FormControl;
+  public password: FormControl;
+  constructor(
+    private authService: AuthService,
+    private router: Router,
+    private alertCtrl: AlertController,
+    private fb: FormBuilder
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: lets call it `formBuilder` not `fb`.

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+  async signIn(email: string, password: string) {
+    const checkEmailVerified = await this.authService.checkUserEmailVerified(email, password);
+    if (checkEmailVerified === undefined) {
+      this.showAlert("User with that email doesn't exist!");
+      return;
+    }
+    if (checkEmailVerified === false) {
+      const alert = await this.alertCtrl
+      .create({
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: Let's implement Internationalisation Service(ngx-translate) and move all the strings in `.json` files

File: ionic-angular-template/src/auth/sign-in/sign-in.component.html [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+                Should at least be 6 characters long.
+              </ion-label>
+            </ion-item>
+          </ion-list>
+        </ion-col>
+      </ion-row>
+      <ion-row class="btn">
+        <ion-col size-sm="6" offset-sm="3">
+          <ion-button
+            type="submit"
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.html]: No need to be "submit" type

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  constructor(
+    private authService: AuthService,
+    private router: Router,
+    private alertCtrl: AlertController,
+    private formBuilder: FormBuilder
+  ) { }
+
+  ngOnInit() {
+    this.signInForm = this.formBuilder.group({
+      email: new FormControl('', Validators.compose([Validators.required, Validators.email])),
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: You can remove "Validator.compose". You can use it directly like: `email: ['', [Validators.required, Validators.email]]`
Compose is used to compose multiple validators into a single function that returns the union of the individual error maps for the provided control.

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+  get email() { 
+    return this.signInForm.get('email'); 
+  }
+
+  get password() { 
+    return this.signInForm.get('password'); 
+  }
+
+  async signIn(email: string, password: string) {
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: You can get the values of `email` and `password` from the getters. This was marked as resolved, but I can still see that is not changed.

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  get email() { 
+    return this.signInForm.get('email'); 
+  }
+
+  get password() { 
+    return this.signInForm.get('password'); 
+  }
+
+  async signIn() {
+    const email = this.signInForm.get('email').value
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: You already have getters for the formControl values. You can use those getters to call them here: this.email.value, this.password.value.

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+  get password() { 
+    return this.signInForm.get('password'); 
+  }
+
+  async signIn() {
+    const email = this.email.value;
+    const password = this.password.value;
+    const checkEmailVerified = await this.authService.checkUserEmailVerified(email, password);
+    if (checkEmailVerified === undefined) {
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: I personally don't like this check. Please review what is returning from the `checkUserEmailVerified` and consider move it in try, catch block.

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+      if (checkEmailVerified === false) {
+        const alert = await this.alertCtrl
+        .create({
+          header: 'Info Message',
+          message: 'If you want to use all functionalities you have to verify your email first!',
+          buttons: ['Okay']
+        })
+        alert.present();
+        }
+      this.authService.signIn().subscribe((d) => {
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: What is `d` ?

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+  get password() { 
+    return this.signInForm.get('password'); 
+  }
+
+  async signIn() {
+    const email = this.email.value;
+    const password = this.password.value;
+    try {
+      const checkEmailVerified = await this.authService.checkUserEmailVerified(email, password);
+      if (checkEmailVerified === false) {
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: In the auth service we are checking: 
`return currentUser.user.emailVerified ? true : false;`
But I think that it is the same as to return just: `currentUser.user.emailVerified`

Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: Let's use `if(! checkEmailVerified)` ?

File: ionic-angular-template/src/auth/sign-in/sign-in.component.html [hunk 18]
```diff
@@ -0,0 +1,62 @@
+<ion-header>
+  <ion-toolbar>
+    <ion-title>Login</ion-title>
+  </ion-toolbar>
+</ion-header>
+
+<ion-content class="ion-padding">
+  <form [formGroup]="signInForm" padding>
+    <ion-grid>
+      <ion-row>
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.html]: The template in Ionic should follow the following rules: 
`<ion-content>` (optional for scrolling the content)
`<ion-grid>
<ion-row>
<ion-col> `
After that you can have the `<form>` tag or `<divs>` or other content.

Comment (IvanShterev) [reply ionic-angular-template/src/auth/sign-in/sign-in.component.html]: I got many errors. At the ionic course is the same as mine. And here [](https://www.positronx.io/ionic-angular-responsive-grid-tutorial-with-examples/) as well.

Comment (IvanShterev) [reply ionic-angular-template/src/auth/sign-in/sign-in.component.html]: `form>
    <ion-grid>
      <ion-row>
        <ion-col size-lg="6" size-md="6" size-sm="6" size="12">
          `....
` </ion-col>
      </ion-row>
    </ion-grid>
  </form>`

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+  async signIn() {
+    const email = this.email.value;
+    const password = this.password.value;
+    try {
+      const checkEmailVerified = await this.authService.checkUserEmailVerified(email, password);
+      if (!checkEmailVerified) {
+        this.showInfoAlert();
+        }
+      this.authService.signIn().subscribe((userData) => {
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: Remove that subscribe. We were discussing to refactor that.

File: ionic-angular-template/src/auth/sign-in/sign-in.component.ts [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+})
+export class SignInComponent implements OnInit {
+  private infoMessages: any;
+  public signInForm: FormGroup;
+  constructor(
+    private authService: AuthService,
+    private router: Router,
+    private alertCtrl: AlertController,
+    private formBuilder: FormBuilder,
+    private tranlate: TranslateService
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.ts]: lets call it: translateService

File: ionic-angular-template/src/auth/sign-in/sign-in.component.scss [hunk 21]
```diff
@@ -0,0 +1,7 @@
...
+    padding: 5%;
```
Comment (ochikov) [inline ionic-angular-template/src/auth/sign-in/sign-in.component.scss]: What happened with the classes from the main stylesheet file, where you have styleForm and etc?

File: ionic-angular-template/src/services/auth.service.ts [hunk 22]
```diff
@@ -33,9 +33,9 @@ export class AuthService {
     });
   }
 
-  signIn() {
-    return this.api.get(`${environment.apiUrl}/users/wallet`);
-}
+  async signIn() {
```
Comment (ochikov) [inline ionic-angular-template/src/services/auth.service.ts]: Let's rename that method ?

**Pull request merged**